unstructured/ingest/interfaces.py

```diff
@@ -670,3 +670,18 @@ def session_handle(self):
     @session_handle.setter
     def session_handle(self, session_handle: BaseSessionHandle):
         self._session_handle = session_handle
+
+
+class WriteConfigSessionHandleMixin(ConfigSessionHandleMixin):
```

Comment:
This is very similar to the dataclass above. For my understanding, why did you not choose dataclass here?

Reply:
![image](https://github.com/Unstructured-IO/unstructured/assets/39141206/7197010f-0fd8-4b19-9185-1f1e78b402d0)

It goes as: 
- step 1

- step 2
  - We said we won't implement per-doc session handlers, but then we will be discarding `IngestDocSessionHandleMixin`s since they are supposed to be used in per-doc operations. And `IngestDocSessionHandleMixin`s have a role of keeping sessions as properties, getting them and setting them. Now we need something else to take that role from `IngestDocSessionHandleMixin`s, and provide it for our batches


Part 1:
  - We already needed to implement a `ConfigSessionHandle` class, which'll implement a `create_session_handle` method via `ConfigSessionHandleMixin`. So we already implemented that part in [here](https://github.com/Unstructured-IO/unstructured/blob/791bf03d00cfca5a0b6e74066b86f5af7c90ad4e/unstructured/ingest/connector/pinecone.py#L43) .
  - The remaining part was having something to keep session handles as a property, and to get & set that property. Normally, IngestDocs do that.

Part 2:
  - Now, in source connectors we perform session related operations per doc (download each doc) , when in destination connectors we perform session related operations per batch (upload each batch).
  - So I didn't want to just choose a random ingest doc within the batch and create a session handle via that random doc. I wanted the design to be obvious that our session handlers are created per batch, rather than a specific doc. That meant we need something else than IngestDocs to get, set and keep our session_handle property.
  - So I thought the `ConfigSessionHandle` class which we've already implemented can keep our session_handle property. No need to create another class.

Part 3:
- Therefore `PineconeWriteConfig` should have inherited both from `ConfigSessionHandleMixin` and `IngestDocSessionHandleMixin`. But `IngestDocSessionHandleMixin` [expects a connector_config](https://github.com/Unstructured-IO/unstructured/blob/f434c3d79faaba78abc4d1bb3d0b7921c3cd70d4/unstructured/ingest/interfaces.py#L662) while PineconeWriteConfig is itself a config! (this would be self reference)
- So I created a new class where the only difference is that create_session_handle is called from the object itself, rather than its connector_config. This way we acknowledge that PineconeWriteConfig (and potentially other destination WriteConfigs in the future) will take the role of both an ingest doc, and a connector config when managing session handles

Reply:
![image](https://github.com/Unstructured-IO/unstructured/assets/39141206/7197010f-0fd8-4b19-9185-1f1e78b402d0)

It goes as: 

Part 1:
  - We already needed to implement a `ConfigSessionHandle` class, which'll implement a `create_session_handle` method via `ConfigSessionHandleMixin`. So we already implemented that part in [here](https://github.com/Unstructured-IO/unstructured/blob/791bf03d00cfca5a0b6e74066b86f5af7c90ad4e/unstructured/ingest/connector/pinecone.py#L43) .
  - The remaining part was having something to keep session handles as a property, and to get & set that property. Normally, IngestDocs do that.

Part 2:
  - Now, in source connectors we perform session related operations per doc (download each doc) , when in destination connectors we perform session related operations per batch (upload each batch).
  - So I didn't want to just choose a random ingest doc within the batch and create a session handle via that random doc. I wanted the design to be obvious that our session handlers are created per batch, rather than a specific doc. That meant we need something else than IngestDocs to get, set and keep our session_handle property.
  - So I thought the `ConfigSessionHandle` class which we've already implemented can keep our session_handle property. No need to create another class.

Part 3:
- Therefore `PineconeWriteConfig` should have inherited both from `ConfigSessionHandleMixin` and `IngestDocSessionHandleMixin`. But `IngestDocSessionHandleMixin` [expects a connector_config](https://github.com/Unstructured-IO/unstructured/blob/f434c3d79faaba78abc4d1bb3d0b7921c3cd70d4/unstructured/ingest/interfaces.py#L662) while PineconeWriteConfig is itself a config! (this would be self reference)
- So I created a new class where the only difference is that create_session_handle is called from the object itself, rather than its connector_config. This way we acknowledge that PineconeWriteConfig (and potentially other destination WriteConfigs in the future) will take the role of both an ingest doc, and a connector config when managing session handles

Reply:
Thanks for this detailed explanation.